cache/demux.go

```diff
@@ -0,0 +1,167 @@
+	}()
+	return d
+}
+
+func (d *demux) add(w *watcher, startingRev int64) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	latestEvent := d.history.PeekLatest()
+
+	// Special case: 0 means “newest”.
+	if startingRev == 0 {
+		if latestEvent != nil {
+			startingRev = latestEvent.Kv.ModRevision + 1
+		} else {
+			d.activeWatchers[w] = 0
+			return
+		}
+	}
+
+	var latestRev int64
+	if latestEvent != nil {
+		latestRev = latestEvent.Kv.ModRevision
+	}
+	if startingRev <= latestRev {
+		d.laggingWatchers[w] = startingRev
+	} else {
+		d.activeWatchers[w] = startingRev
+	}
+}
+
+func (d *demux) remove(w *watcher) {
+	func() {
+		d.mu.Lock()
+		defer d.mu.Unlock()
+		delete(d.activeWatchers, w)
+		delete(d.laggingWatchers, w)
+	}()
+	w.Stop()
+}
+
+func (d *demux) broadcast(event *clientv3.Event) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	windowSize := len(d.history.buffer)
+	for w, nextRev := range d.laggingWatchers {
+		if int(event.Kv.ModRevision-nextRev+1) >= windowSize {
+			w.Stop()
+			delete(d.laggingWatchers, w)
+		}
+	}
+
+	for w, nextRev := range d.activeWatchers {
+		if event.Kv.ModRevision < nextRev {
+			continue
+		}
+		if !w.enqueueEvent(event) { // buffer overflow
+			gap := int(event.Kv.ModRevision - nextRev + 1)
+			if gap >= windowSize {
+				w.Stop()
+			} else {
+				d.laggingWatchers[w] = nextRev
+			}
+			delete(d.activeWatchers, w)
+		} else {
+			d.activeWatchers[w] = event.Kv.ModRevision + 1
+		}
+	}
+}
+
+// Called when etcd compaction invalidates our cached history, so clients should resubscribe.
+func (d *demux) purgeAll() {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	for w := range d.activeWatchers {
+		w.Stop()
+	}
+	for w := range d.laggingWatchers {
+		w.Stop()
+	}
+	d.activeWatchers = make(map[*watcher]int64)
+	d.laggingWatchers = make(map[*watcher]int64)
+}
+
+// resyncLoop periodically tries to catch lagging watchers up by replaying events from History.
+func (d *demux) resyncLoop(ctx context.Context) {
+	timer := time.NewTimer(d.resyncInterval)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			d.resyncLaggingWatchers()
+			timer.Reset(d.resyncInterval)
+		}
+	}
+}
+
+func (d *demux) resyncLaggingWatchers() {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	for w, nextRev := range d.laggingWatchers {
+		// TODO: re-enable key‐predicate in Filter when non‐zero startRev or performance tuning is needed
+		missedEvents, oldestRev, _ := d.history.Filter(AfterRev(nextRev)) // TODO: use binary search on the ring buffer to locate the first entry >= nextRev instead of a full scan
+		if oldestRev != 0 && nextRev < oldestRev {
+			w.Stop()
+			delete(d.laggingWatchers, w)
+			continue
+		}
+
+		var lastSent int64
+		for _, event := range missedEvents {
+			if !w.enqueueEvent(event) { // buffer overflow: watcher still lagging
+				break
+			}
+			lastSent = event.Kv.ModRevision
+		}
+
+		if lastSent == 0 {
+			d.laggingWatchers[w] = nextRev
+			continue
+		}
+
+		nextRev = lastSent + 1
```

Comment:
Don't see reason for lastSent. 
```suggestion
		for _, event := range missedEvents {
			if !w.enqueueEvent(event) { // buffer overflow: watcher still lagging
				break
			}
			nextRev = event.Kv.ModRevision + 1
		}
```